react/features/base/redux/StateListenerRegistry.js

```diff
@@ -0,0 +1,159 @@
+ * @param {any} prevSelection - The value previously derived from the redux
+ * store/state by the associated {@code Selector}. The {@code Listener} is
+ * invoked only if {@code prevSelection} and {@code selection} are different.
+ * Immutable!
+ */
+type Listener = (selection: any, store: Store, prevSelection: any) => void;
+
+/**
+ * The type selector supported for registration with
+ * {@link StateListenerRegistry} in association with a {@link Listener}.
+ *
+ * @param {Object} state - The redux state from which the {@code Selector} is to
+ * derive data.
+ * @param {any} prevSelection - The value previously derived from the redux
+ * store/state by the {@code Selector}. Provided in case the {@code Selector}
+ * needs to derive the returned value from the specified {@code state} and
+ * {@code prevSelection}. Immutable!
+ * @returns {any} The value derived from the specified {@code state} and/or
+ * {@code prevSelection}. The associated {@code Listener} will only be invoked
+ * if the returned value is other than {@code prevSelection}.
+ */
+type Selector = (state: Object, prevSelection: any) => any;
+
+/**
+ * A type of a {@link Selector}-{@link Listener} association in which the
+ * {@code Listener} listens to changes in the values derived from a redux
+ * store/state by the {@code Selector}.
+ */
+type SelectorListener = {
+
+    /**
+     * The {@code Listener} which listens to changes in the values selected by
+     * {@link selector}.
+     */
+    listener: Listener,
+
+    /**
+     * The {@code Selector} which selects values whose changes are listened to
+     * by {@link listener}.
+     */
+    selector: Selector
+};
+
+/**
+ * A registry listeners which listen to changes in a redux store/state.
+ */
+class StateListenerRegistry {
+    /**
+     * The {@link Listener}s registered with this {@code StateListenerRegistry}
+     * to be notified when the values derived by associated {@link Selector}s
+     * from a redux store/state change.
+     */
+    _selectorListeners: Set<SelectorListener> = new Set();
+
+    _listener: (Store) => void;
+
+    /**
+     * Invoked by a specific redux store any time an action is dispatched, and
+     * some part of the state (tree) may potentially have changed.
+     *
+     * @param {Object} context - The redux store invoking the listener and the
+     * private state of this {@code StateListenerRegistry} associated with the
+     * redux store.
+     * @returns {void}
+     */
+    _listener({ prevSelections, store }: {
+            prevSelections: Map<SelectorListener, any>,
+            store: Store
+    }) {
+        for (const selectorListener of this._selectorListeners) {
+            const prevSelection = prevSelections.get(selectorListener);
+
+            try {
+                const selection
+                    = selectorListener.selector(
+                        store.getState(),
+                        prevSelection);
+
+                if (prevSelection !== selection) {
```

Comment:
I don't think this is enough for anything other than simple objects. How about `_.isEqual` instead?

Reply:
Because it's easier. Though I haven't added shallow equality, there is still complete flexibility though:

* If a Selector needs to return a simple value, obviously, `===`/`!==` will suffice.
* If the Selector needs to return a combination of simple values, they should rather compute the next/new `selection` based on the `prevSelection`. For example, with the help of our base/redux functions `set` or `assign`, `prevSelection` can immutably be transformed into new/next `selection`.

Anyway, I'm against `_.isEqual` in particular because it's deep equality which is very different from what redux and react-redux give us.

Reply:
Ack.